### ISSUE TITLE
Thoughts for improving the boxplot for categorical EDA

### ISSUE THREAD

**Issue by ianozsvald** [dabl/dabl#179]
`plot` has to make assumptions about the data, I think the categorical plot could be improved to help with low-frequency category cases. The following plots come from the autompg dataset. The `cylinder` column has 2 low-frequency categories (3 & 5) and 3 high-frequency categories. Notebook:
https://github.com/ianozsvald/dabl_experiments/blob/master/autompg_example.ipynb

Dabl `plot` shows 5 categories, out of sequence, all categories appear to be 'good':
![image](https://user-images.githubusercontent.com/273210/73447179-1a8e4000-4356-11ea-991b-5f46d5c312c2.png)

The above plot seems to use median-sorted display order, it makes it look as though there's a sequence in the data which isn't true [8, 6, 3, 5, 4] for cylinders.

Adding a countplot shows low-frequency for 2 categories and adding `notch=True` shows very wide confidence intervals around the median for these - this looks ugly for those unfamiliar with CIs but is informative if you know it means 'wide CI, probably caused by few datapoints':
![image](https://user-images.githubusercontent.com/273210/73447375-7fe23100-4356-11ea-907d-7e470fddc362.png)

A neater plot would make the countplot a smaller part of the overall plot but it still takes up a lot of space:
![image](https://user-images.githubusercontent.com/273210/73447432-9be5d280-4356-11ea-91f1-3a83255ad385.png)

Adding frequency counts into the y-axis labels might be a useful way to summarise this:
![image](https://user-images.githubusercontent.com/273210/73447513-cb94da80-4356-11ea-83bf-9b7e8631df57.png)

The y-axis label could say something coarser like "<10", "<1k", "<10", "<100", "<1k" or somesuch to avoid too much detail.

I guess one question is - what's the purpose of dabl for EDA? Is it to give fine-grained results that have been automatically interpreted to tell a story? Is it to give the user a _steer_ towards interesting relationships which they should examine in further depth? I think that notches and countplots and default sorting decisions will always be righter or wronger for different cases.

**Comment by amueller** [comment]
I think it is to steer towards interesting relationships. We should make good choices whenever we can. I definitely agree with the ordering, its something that was on my radar. Right now it's basically just a feature that's not implemented because there is no clear distinction between ordinal and categorical for plotting.

**Comment by amueller** [comment]
I think adding notches and adding counts to the ticks would be good. The problem with adding counts in a histogram is that this means we have to use multiple axes per plot. That's tricky, which is why seaborn basically doesn't allow it for grids. I hacked around that in the class histograms by basically reimplementing histograms. We could also change the height of the boxes but I don't think that's common practice, right?

I have to admit the regression plots are much less developed than the classification ones (though the same is true for classification and mosaic plot)

**Comment by ianozsvald** [comment]
I think adding the notches isn't bad (as long as we document the "bad shape" for wide CIs), I think  an approximate count in the tick labels is probably reasonable for larger counts. Agreed that having a pair of subplots is probably overkill and having variable-height bars feels weird.
I can raise a PR for the above if you fancy?

**Comment by amueller** [comment]
Yes please! With you as contributor, dabl will finally be taken seriously ;)

**Comment by ianozsvald** [comment]
Hey there. Here's a first draft - alphanumeric sort on the labels with counts in brackets:
![image](https://user-images.githubusercontent.com/273210/75590791-0da85d80-5a76-11ea-8e9a-be1766600551.png)
Counts are verbatim for <100, otherwise they're marked >100, >1K, >10K (for all larger counts). The lack of notches hides the wider uncertainty for the median _but_ having the counts in brackets should give the user a good clue about areas of concern. Thoughts?

**Comment by amueller** [comment]
Thanks, that's cool! Do you think using an alphanumeric sorting is always good? My idea was to only use it on ordinal data. Maybe as a first approximation only on integers?

**Comment by ianozsvald** [comment]
I used `order = sorted(np.unique(X[col]))` as a first crack, so it sorts any label type. Without a sort then text labels will be in natural-discovery order - I'm not sure that that's "better" or just different?  
Also it looks like inside `plot_regression_categorical` the `X` dataframe already has categorical encoding for `territory` and `cylinders` so I can't tell them apart. The `df` I pass in has `string` and `float64` respectively so I'm guessing you recode this somewhere?
I'm happy to play (check for categories containing int/float values?), give me a steer

**Comment by amueller** [comment]
this is discussed in #202 so it's not super obvious unfortunately. Right now it's happening in ``plot``:
https://github.com/dabl/dabl/blob/master/dabl/plot/supervised.py#L551
and
https://github.com/dabl/dabl/blob/master/dabl/plot/supervised.py#L557


We could hack it and just check if all the categories are integers, that's probably a reasonable heuristic.

Right now, all categories are ordered by their median value:
https://github.com/dabl/dabl/blob/master/dabl/plot/supervised.py#L152

I think that's a good idea if the categories are not integers because it allows you to compare two similar categories more easily (they will be next to each other). That was at least the motivation when I wrote this sorting logic.

**Comment by ianozsvald** [comment]
I got caught up with NaNs last night (`sorted` doesn't like a mix of strings and NaNs!) and took a break. In your Ames Housing Dataset the regression categorical plot doesn't show NaNs and yet they're the most abundant:

```
>>> ames_df['Alley'].value_counts(dropna=False)
NaN     2732
Grvl     120
Pave      78
```

Am I right thinking that dropping NaNs is your preference? I wonder if reporting them in some way would be important (or is that perhaps the job of a tool like Pandas Profiling)?

Re. your previous point about two similar categories of string-labels being sorted by median - yes, nice observation.

**Comment by ianozsvald** [comment]
Ok, I'm getting closer. Here's the autompg categoricals, the first is ordinal (sorted numerically), the second is not (sorted by median):
![image](https://user-images.githubusercontent.com/273210/75608250-ccf02900-5af5-11ea-9e4a-45dee901c8cf.png)

Here's the same for the Ames housing (which contains the NaNs that caused some confusion):
![image](https://user-images.githubusercontent.com/273210/75608267-e7c29d80-5af5-11ea-8fde-ca3140d609a7.png)

Here's your original Ames for comparison: https://github.com/dabl/dabl/blob/master/demo.ipynb

Note that I've lost the x-axis labels (will fix), everything else I think is looking sensible. Thoughts?
